### PR TITLE
Add redundant nil coalesing operator opt-in rule

### PR DESCRIPTION
**Pull request by daniel-beard**
- https://github.com/realm/SwiftLint/issues/764

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -8,7 +8,9 @@
 
 ##### Enhancements
 
-* None.
+* Add `RedundantNilCoalesingRule` Opt-In rule that warns against `?? nil`.
```
Comment (jpsim) [inline CHANGELOG.md]: Missing two trailing spaces, as noted in contributing.md

Also, this should be moved to a new top-level changelog section now that 0.12.0 is out.

File: Source/SwiftLintFramework/Extensions/File+Cache.swift [hunk 2]
```diff
@@ -86,6 +86,7 @@ extension File {
 
     internal var assertHandler: (() -> ())? {
         get {
+            // swiftlint:disable:next redundant_nil_coalesing_operator
```
Comment (jpsim) [inline Source/SwiftLintFramework/Extensions/File+Cache.swift]: I don't think we should turn this rule on for SwiftLint.

File: Source/SwiftLintFramework/Rules/RedundantNilCoalesingRule.swift [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+    public var configuration = SeverityConfiguration(.Warning)
+
+    public init() {}
+
+    public static let description = RuleDescription(
+        identifier: "redundant_nil_coalesing_operator",
+        name: "Redundant Nil Coalesing Operator",
+        description: "nil coalescing operator is only evaluated if the lhs is nil " +
+        ", coalesing operator with nil as rhs is redundant",
```
Comment (jpsim) [inline Source/SwiftLintFramework/Rules/RedundantNilCoalesingRule.swift]: could you please indent this one level?

File: Source/SwiftLintFramework/Rules/RedundantNilCoalesingRule.swift [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+}
+
+public struct RedundantNilCoalesingRule: OptInRule, ConfigurationProviderRule {
+
+    public var configuration = SeverityConfiguration(.Warning)
+
+    public init() {}
+
+    public static let description = RuleDescription(
+        identifier: "redundant_nil_coalesing_operator",
```
Comment (jpsim) [inline Source/SwiftLintFramework/Rules/RedundantNilCoalesingRule.swift]: I'd drop "operator" from this (and the name): `redundant_nil_coalesing`

File: Source/SwiftLintFramework/Rules/RedundantNilCoalesingRule.swift [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+        name: "Redundant Nil Coalesing Operator",
```
Comment (jpsim) [inline Source/SwiftLintFramework/Rules/RedundantNilCoalesingRule.swift]: I'd drop "Operator" from this: "Redundant Nil Coalesing"

**Comment by jpsim** [comment]
I feel like this could easily conform to `CorrectableRule`.

**Comment by daniel-beard** [comment]
@jpsim updated

**Comment by daniel-beard** [comment]
Not sure why the build is stuck...

**Comment by jpsim** [comment]
The build is stuck because Travis isn't running OS X jobs at this time:

> We've identified issues with multiple OSX build hosts and we're working with our infrastructure provider to resolve this, as well as performing some maintenance tasks to help improve stability. Builds will continue to be paused while we work through this. 

https://www.traviscistatus.com/

File: CHANGELOG.md [hunk 6]
```diff
@@ -1,3 +1,19 @@
...
+
+##### Breaking
+
+* None.
+
+##### Enhancements
+
+* Add `RedundantNilCoalesingRule` Opt-In rule that warns against `?? nil`.
+  [Daniel Beard](https://github.com/daniel-beard)
+  [#764](https://github.com/realm/SwiftLint/issues/764)
```
Comment (jpsim) [inline CHANGELOG.md]: trailing whitespace is at the wrong place. see other changelog entries for format.

**Comment by jpsim** [comment]
This is missing a test in `RulesTests.swift`. Otherwise looks good, but I'll want to wait for Travis to be back online.

File: CHANGELOG.md (hunk 6 above)
Comment (daniel-beard) [inline CHANGELOG.md]: got it, fixed.

**Comment by daniel-beard** [comment]
Added test as well.

**Comment by jpsim** [comment]
👍 local test passes so merging.

**Pull request merged**